### PR TITLE
Move forced-id to HFJ_RESOURCE - step 2

### PR DESCRIPTION
**Pull request by michaelabuckley**
Start using new fhir_id column in hfj_resource instead for queries.  Still writes forced_id to support zero-downtime.

Reviewer notes:
The existing test coverage seems pretty good, so I'm not too worried about the core behaviour.  Two big worries are:
- the custom migration - it is tested on Sql Server, H2, PG, Oracle, and MySql.  Haven't tested Derby.
- concurrent access by old clients.

We still write to hfj_forced_id, but never query it now.  Older servers should have been populating the new FHIR_ID column, and the migration fills in the rest and makes it not-null and indexed.

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/custom/IResourceTableDaoImpl.java [hunk 1]
```diff
@@ -27,9 +27,12 @@
...
 @Component
-// Don't change the name of this class.  Spring Data requires the name to match.
-// See https://stackoverflow.com/questions/11880924/how-to-add-custom-method-to-spring-data-jpa
-public class IForcedIdDaoImpl implements IForcedIdQueries {
+/**
+ * Custom query implementations.
+ * Don't change the name of this class.  Spring Data requires the name to match.
+ * https://stackoverflow.com/questions/11880924/how-to-add-custom-method-to-spring-data-jpa
+ */
+public class IResourceTableDaoImpl implements IForcedIdQueries {
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/custom/IResourceTableDaoImpl.java:35]: question: I imagine you have taken the "Dont change the name of this class" comment into account, when you changed the name, and consider this safe to do.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/custom/IResourceTableDaoImpl.java [hunk 2]
```diff
@@ -78,11 +79,9 @@ public Collection<Object[]> findAndResolveByForcedIdWithNoTypeInPartition(
 			Collection<String> theForcedIds,
 			Collection<Integer> thePartitionId,
 			boolean theExcludeDeleted) {
-		String query = "" + "SELECT "
-				+ "   f.myResourceType, f.myResourcePid, f.myForcedId, t.myDeleted "
-				+ "FROM ForcedId f "
-				+ "JOIN ResourceTable t ON t.myId = f.myResourcePid "
-				+ "WHERE f.myResourceType = :resource_type AND f.myForcedId IN ( :forced_id ) AND f.myPartitionIdValue IN ( :partition_id )";
+		String query = "SELECT t.myResourceType, t.id, t.myFhirId, t.myDeleted "
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/custom/IResourceTableDaoImpl.java:82]: question: is `t.id` correct here? I am scanning that class and don't see that field.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/TermReadSvcImpl.java [hunk 3]
```diff
@@ -2972,14 +2971,15 @@ public Optional<TermValueSet> findCurrentTermValueSet(String theUrl) {
 	public Optional<IBaseResource> readCodeSystemByForcedId(String theForcedId) {
 		@SuppressWarnings("unchecked")
 		List<ResourceTable> resultList = (List<ResourceTable>) myEntityManager
-				.createQuery("select f.myResource from ForcedId f "
-						+ "where f.myResourceType = 'CodeSystem' and f.myForcedId = '" + theForcedId + "'")
+				.createQuery("select r from ResourceTable r "
+						+ "where r.myResourceType = 'CodeSystem' and r.myFhirId = :fhirId")
+				.setParameter("fhirId", theForcedId)
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/term/TermReadSvcImpl.java:2976]: nice catch

File: hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/ForceIdMigrationCopyTask.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+			ourLog.info("Migrating client-assigned ids for pids: {}-{}", batchStart, batchEnd);
+
+			// This should be fast-ish since fhir_id isn't indexed yet,
+			// and we're walking both hfj_resource and hfj_forced_id in insertion order.
+			executeSql(
+					"hfj_resource",
+					"update hfj_resource " + "set fhir_id = coalesce( "
+							+ // use first non-null value: forced_id if present, otherwise res_id
+							"   (select f.forced_id from hfj_forced_id f where f.resource_pid = res_id), "
+							+ "   cast(res_id as char(64)) "
```
Comment (tadgh) [inline hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/ForceIdMigrationCopyTask.java:53]: Question: have you tested this particular SQL against all our DB platforms?

**Review by tadgh — APPROVED** [review]

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/custom/IResourceTableDaoImpl.java (hunk 1 above)
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/custom/IResourceTableDaoImpl.java:35]: Yes.  The comment is about the `I` at the front.  The actual queries did move from ForcedDao to ResourceTable.

File: hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/ForceIdMigrationCopyTask.java (hunk 4 above)
Comment (michaelabuckley) [reply hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/ForceIdMigrationCopyTask.java:53]: Manually against Sql Server, H2, PG, Oracle, and MySql. Haven't tested Derby.

Comment (michaelabuckley) [reply hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/ForceIdMigrationCopyTask.java:53]: I figure nobody is running prod on Derby.  It's pretty standard sql.

**Pull request merged**